File: finanze/infrastructure/repository/transaction/transaction_repository.py
from datetime import datetime
from typing import List, Optional, Set
from uuid import UUID

from application.ports.transaction_port import TransactionPort
from dateutil.tz import tzlocal
from domain.dezimal import Dezimal
from domain.entity import Entity
from domain.fetch_record import DataSource
from domain.global_position import EquityType, FundType, ProductType
from domain.transactions import (
    AccountTx,
    BaseInvestmentTx,
    BaseTx,
    CryptoCurrencyTx,
    DepositTx,
    FactoringTx,
    FundPortfolioTx,
    FundTx,
    RealEstateCFTx,
    StockTx,
    TransactionQueryRequest,
    Transactions,
    TxType,
)
from infrastructure.repository.db.client import DBClient
from infrastructure.repository.transaction.queries import TransactionQueries


def _map_account_row(row) -> AccountTx:
    entity = Entity(
        id=UUID(row["entity_id"]),
        name=row["entity_name"],
        natural_id=row["entity_natural_id"],
        type=row["entity_type"],
        origin=row["entity_origin"],
        icon_url=row["icon_url"],
    )

    return AccountTx(
        id=UUID(row["id"]),
        ref=row["ref"],
        name=row["name"],
        amount=Dezimal(row["amount"]),
        currency=row["currency"],
        type=TxType(row["type"]),
        date=datetime.fromisoformat(row["date"]),
        entity=entity,
        source=DataSource(row["source"]),
        product_type=ProductType.ACCOUNT,
        fees=Dezimal(row["fees"]),
        retentions=Dezimal(row["retentions"]),
        interest_rate=Dezimal(row["interest_rate"]) if row["interest_rate"] else None,
        avg_balance=Dezimal(row["avg_balance"]) if row["avg_balance"] else None,
        net_amount=Dezimal(row["net_amount"]) if row["net_amount"] else None,
    )


def _map_investment_row(
    row, fallback_entity: Optional[Entity] = None
) -> BaseInvestmentTx:
    entity = (
        Entity(
            id=UUID(row["entity_id"]),
            name=row["entity_name"],
            natural_id=row["entity_natural_id"],
            type=row["entity_type"],
            origin=row["entity_origin"],
            icon_url=row["icon_url"],
        )
        if row["entity_id"]
        else fallback_entity
    )

    common = {
        "id": UUID(row["id"]),
        "ref": row["ref"],
        "name": row["name"],
        "amount": Dezimal(row["amount"]),
        "currency": row["currency"],
        "type": TxType(row["type"]),
        "date": datetime.fromisoformat(row["date"]),
        "entity": entity,
        "source": DataSource(row["source"]),
        "product_type": ProductType(row["product_type"]),
    }

    if row["product_type"] == ProductType.STOCK_ETF.value:
        return StockTx(
            **common,
            isin=row["isin"] if row["isin"] else None,
            ticker=row["ticker"],
            market=row["market"],
            shares=Dezimal(row["shares"]),
            price=Dezimal(row["price"]),
            net_amount=Dezimal(row["net_amount"]) if row["net_amount"] else None,
            fees=Dezimal(row["fees"]),
            retentions=Dezimal(row["retentions"]) if row["retentions"] else None,
            order_date=datetime.fromisoformat(row["order_date"])
            if row["order_date"]
            else None,
            linked_tx=row["linked_tx"],
            equity_type=EquityType(row["product_subtype"])
            if row["product_subtype"]
            else None,
        )
    elif row["product_type"] == ProductType.CRYPTO.value:
        return CryptoCurrencyTx(
            **common,
            symbol=row["ticker"],
            currency_amount=Dezimal(row["shares"]),
            price=Dezimal(row["price"]),
            net_amount=Dezimal(row["net_amount"]) if row["net_amount"] else None,
            fees=Dezimal(row["fees"]),
            retentions=Dezimal(row["retentions"]) if row["retentions"] else None,
            order_date=datetime.fromisoformat(row["order_date"])
            if row["order_date"]
            else None,
            contract_address=row["asset_contract_address"],
        )
    elif row["product_type"] == ProductType.FUND.value:
        return FundTx(
            **common,
            isin=row["isin"] if row["isin"] else None,
            market=row["market"],
            shares=Dezimal(row["shares"]),
            price=Dezimal(row["price"]),
            net_amount=Dezimal(row["net_amount"]),
            fees=Dezimal(row["fees"]),
            retentions=Dezimal(row["retentions"]) if row["retentions"] else None,
            order_date=datetime.fromisoformat(row["order_date"])
            if row["order_date"]
            else None,
            fund_type=FundType(row["product_subtype"])
            if row["product_subtype"]
            else None,
        )
    elif row["product_type"] == ProductType.FUND_PORTFOLIO.value:
        return FundPortfolioTx(
            **common,
            fees=Dezimal(row["fees"]),
            portfolio_name=row["portfolio_name"] if row["portfolio_name"] else None,
            iban=row["iban"] if row["iban"] else None,
        )
    elif row["product_type"] == ProductType.FACTORING.value:
        return FactoringTx(
            **common,
            net_amount=Dezimal(row["net_amount"]),
            fees=Dezimal(row["fees"]),
            retentions=Dezimal(row["retentions"]),
        )
    elif row["product_type"] == ProductType.REAL_ESTATE_CF.value:
        return RealEstateCFTx(
            **common,
            net_amount=Dezimal(row["net_amount"]),
            fees=Dezimal(row["fees"]),
            retentions=Dezimal(row["retentions"]),
        )
    elif row["product_type"] == ProductType.DEPOSIT.value:
        return DepositTx(
            **common,
            net_amount=Dezimal(row["net_amount"]),
            fees=Dezimal(row["fees"]),
            retentions=Dezimal(row["retentions"]),
        )
    else:
        raise ValueError(f"Unknown product type: {row['product_type']}")


class TransactionSQLRepository(TransactionPort):
    def __init__(self, client: DBClient):
        self._db_client = client

    def save(self, data: Transactions):
        if data.investment:
            self._save_investment(data.investment)
        if data.account:
            self._save_account(data.account)

    def _save_investment(self, txs: List[BaseInvestmentTx]):
        with self._db_client.tx() as cursor:
            for tx in txs:
                entry = {
                    "id": str(tx.id),
                    "ref": tx.ref,
                    "name": tx.name,
                    "amount": str(tx.amount),
                    "currency": tx.currency,
                    "type": tx.type.value,
                    "date": tx.date.isoformat(),
                    "entity_id": str(tx.entity.id),
                    "is_real": tx.source == DataSource.REAL,
                    "source": tx.source.value,
                    "product_type": tx.product_type.value,
                    "created_at": datetime.now(tzlocal()).isoformat(),
                    "isin": None,
                    "ticker": None,
                    "market": None,
                    "shares": None,
                    "price": None,
                    "net_amount": None,
                    "fees": None,
                    "retentions": None,
                    "order_date": None,
                    "linked_tx": None,
                    "interests": None,
                    "iban": None,
                    "portfolio_name": None,
                    "product_subtype": None,
                    "asset_contract_address": None,
                }

                if isinstance(tx, StockTx):
                    entry.update(
                        {
                            "isin": tx.isin,
                            "ticker": tx.ticker,
                            "market": tx.market,
                            "shares": str(tx.shares),
                            "price": str(tx.price),
                            "net_amount": str(tx.net_amount),
                            "fees": str(tx.fees),
                            "retentions": str(tx.retentions) if tx.retentions else None,
                            "order_date": tx.order_date.isoformat()
                            if tx.order_date
                            else None,
                            "linked_tx": tx.linked_tx,
                            "product_subtype": tx.equity_type.value
                            if tx.equity_type
                            else None,
                        }
                    )
                elif isinstance(tx, CryptoCurrencyTx):
                    entry.update(
                        {
                            "ticker": tx.symbol,
                            "shares": str(tx.currency_amount),
                            "price": str(tx.price),
                            "net_amount": str(tx.net_amount),
                            "fees": str(tx.fees),
                            "retentions": str(tx.retentions) if tx.retentions else None,
                            "order_date": tx.order_date.isoformat()
                            if tx.order_date
                            else None,
                            "asset_contract_address": tx.contract_address,
                        }
                    )
                elif isinstance(tx, FundTx):
                    entry.update(
                        {
                            "isin": tx.isin,
                            "market": tx.market,
                            "shares": str(tx.shares),
                            "price": str(tx.price),
                            "net_amount": str(tx.net_amount),
                            "fees": str(tx.fees),
                            "retentions": str(tx.retentions) if tx.retentions else None,
                            "order_date": tx.order_date.isoformat()
                            if tx.order_date
                            else None,
                            "product_subtype": tx.fund_type.value
                            if tx.fund_type
                            else None,
                        }
                    )
                elif isinstance(tx, FundPortfolioTx):
                    entry.update(
                        {
                            "fees": str(tx.fees),
                            "portfolio_name": tx.portfolio_name,
                            "iban": str(tx.iban) if tx.iban else None,
                        }
                    )
                elif isinstance(tx, (FactoringTx, RealEstateCFTx, DepositTx)):
                    entry.update(
                        {
                            "net_amount": str(tx.net_amount),
                            "fees": str(tx.fees),
                            "retentions": str(tx.retentions),
                        }
                    )

                cursor.execute(
                    TransactionQueries.INSERT_INVESTMENT,
                    entry,
                )

    def _save_account(self, txs: List[AccountTx]):
        with self._db_client.tx() as cursor:
            for tx in txs:
                cursor.execute(
                    TransactionQueries.INSERT_ACCOUNT,
                    (
                        str(tx.id),
                        tx.ref,
                        tx.name,
                        str(tx.amount),
                        tx.currency,
                        tx.type.value,
                        tx.date.isoformat(),
                        str(tx.entity.id),
                        tx.source == DataSource.REAL,
                        tx.source.value,
                        datetime.now(tzlocal()).isoformat(),
                        str(tx.fees),
                        str(tx.retentions),
                        str(tx.interest_rate) if tx.interest_rate else None,
                        str(tx.avg_balance) if tx.avg_balance else None,
                        str(tx.net_amount) if tx.net_amount else None,
                    ),
                )

    def get_all(
        self,
        real: Optional[bool] = None,
        excluded_entities: Optional[list[UUID]] = None,
    ) -> Transactions:
        return Transactions(
            investment=self._get_investment_txs(real, excluded_entities),
            account=self._get_account_txs(real, excluded_entities),
        )

    def _get_investment_txs(
        self,
        real: Optional[bool] = None,
        excluded_entities: Optional[list[UUID]] = None,
    ) -> List[BaseInvestmentTx]:
        with self._db_client.read() as cursor:
            params: list[str] = []
            query = TransactionQueries.INVESTMENT_SELECT_BASE.value

            conditions: list[str] = []
            if real is not None:
                if real:
                    conditions.append("it.source = 'REAL'")
                else:
                    conditions.append("it.source IN ('MANUAL', 'SHEETS')")

            if excluded_entities:
                placeholders = ", ".join("?" for _ in excluded_entities)
                conditions.append(f"it.entity_id NOT IN ({placeholders})")
                params.extend([str(e) for e in excluded_entities])

            if conditions:
                query += " WHERE " + " AND ".join(conditions)

            query += " ORDER BY it.date ASC"
            cursor.execute(query, tuple(params))

            return [_map_investment_row(row) for row in cursor.fetchall()]

    def _get_account_txs(
        self,
        real: Optional[bool] = None,
        excluded_entities: Optional[list[UUID]] = None,
    ) -> List[AccountTx]:
        with self._db_client.read() as cursor:
            params: list[str] = []
            query = TransactionQueries.ACCOUNT_SELECT_BASE.value

            conditions: list[str] = []
            if real is not None:
                if real:
                    conditions.append("at.source = 'REAL'")
                else:
                    conditions.append("at.source IN ('MANUAL', 'SHEETS')")

            if excluded_entities:
                placeholders = ", ".join("?" for _ in excluded_entities)
                conditions.append(f"at.entity_id NOT IN ({placeholders})")
                params.extend([str(e) for e in excluded_entities])

            if conditions:
                query += " WHERE " + " AND ".join(conditions)

            query += " ORDER BY at.date ASC"

            cursor.execute(query, tuple(params))
            return [_map_account_row(row) for row in cursor.fetchall()]

    def _get_investment_txs_by_entity(self, entity_id: UUID) -> List[BaseInvestmentTx]:
        with self._db_client.read() as cursor:
            cursor.execute(
                TransactionQueries.INVESTMENT_SELECT_BY_ENTITY,
                (str(entity_id),),
            )
            return [_map_investment_row(row) for row in cursor.fetchall()]

    def _get_account_txs_by_entity(self, entity_id: UUID) -> List[AccountTx]:
        with self._db_client.read() as cursor:
            cursor.execute(
                TransactionQueries.ACCOUNT_SELECT_BY_ENTITY,
                (str(entity_id),),
            )
            return [_map_account_row(row) for row in cursor.fetchall()]

    def get_refs_by_entity(self, entity_id: UUID) -> Set[str]:
        with self._db_client.read() as cursor:
            cursor.execute(
                TransactionQueries.GET_REFS_BY_ENTITY,
                (str(entity_id), str(entity_id)),
            )
            return {row[0] for row in cursor.fetchall()}

    def get_by_entity(self, entity_id: UUID) -> Transactions:
        return Transactions(
            investment=self._get_investment_txs_by_entity(entity_id),
            account=self._get_account_txs_by_entity(entity_id),
        )

    def get_by_entity_and_source(
        self, entity_id: UUID, source: DataSource
    ) -> Transactions:
        with self._db_client.read() as cursor:
            cursor.execute(
                TransactionQueries.INVESTMENT_AND_ACCOUNT_BY_ENTITY_AND_SOURCE,
                (str(entity_id), source.value),
            )
            investment = [_map_investment_row(row) for row in cursor.fetchall()]

            cursor.execute(
                TransactionQueries.ACCOUNT_BY_ENTITY_AND_SOURCE,
                (str(entity_id), source.value),
            )
            account = [_map_account_row(row) for row in cursor.fetchall()]

        return Transactions(investment=investment, account=account)

    def get_refs_by_source_type(self, real: bool) -> Set[str]:
        with self._db_client.read() as cursor:
            cursor.execute(
                TransactionQueries.GET_REFS_BY_SOURCE_TYPE,
                (real, real),
            )
            return {row[0] for row in cursor.fetchall()}

    def get_by_filters(self, query: TransactionQueryRequest) -> list[BaseTx]:
        params = []
        base_sql = TransactionQueries.GET_BY_FILTERS_BASE.value

        conditions = []
        if query.entities:
            placeholders = ", ".join("?" for _ in query.entities)
            conditions.append(f"tx.entity_id IN ({placeholders})")
            params.extend([str(e) for e in query.entities])
        if query.excluded_entities:
            placeholders = ", ".join("?" for _ in query.excluded_entities)
            conditions.append(
                f"(tx.entity_id NOT IN ({placeholders}) OR tx.is_real = FALSE)"
            )
            params.extend([str(e) for e in query.excluded_entities])
        if query.product_types:
            placeholders = ", ".join("?" for _ in query.product_types)
            conditions.append(f"tx.product_type IN ({placeholders})")
            params.extend([pt.value for pt in query.product_types])
        if query.types:
            placeholders = ", ".join("?" for _ in query.types)
            conditions.append(f"tx.type IN ({placeholders})")
            params.extend([t.value for t in query.types])
        if query.from_date:
            conditions.append("tx.date >= ?")
            params.append(query.from_date.isoformat())
        if query.to_date:
            conditions.append("tx.date <= ?")
            params.append(query.to_date.isoformat())
        if query.historic_entry_id:
            conditions.append(
                "EXISTS (SELECT 1 FROM investment_historic_txs ht WHERE ht.tx_id = tx.id AND ht.historic_entry_id = ?)"
            )
            params.append(str(query.historic_entry_id))

        where_clause = f"WHERE {' AND '.join(conditions)}" if conditions else ""
        order_pagination = "ORDER BY tx.date DESC LIMIT ? OFFSET ?"
        offset = (query.page - 1) * query.limit
        params.extend([query.limit, offset])
        sql = f"{base_sql} {where_clause} {order_pagination}"

        with self._db_client.read() as cursor:
            cursor.execute(sql, tuple(params))
            rows = cursor.fetchall()

        tx_list = []
        for row in rows:
            if row["product_type"] == "ACCOUNT":
                tx = _map_account_row(row)
            else:
                tx = _map_investment_row(row)
            tx_list.append(tx)

        return tx_list

    def delete_by_source(self, source: DataSource):
        with self._db_client.tx() as cursor:
            cursor.execute(
                TransactionQueries.DELETE_INVESTMENT_BY_SOURCE,
                (source,),
            )
            cursor.execute(
                TransactionQueries.DELETE_ACCOUNT_BY_SOURCE,
                (source,),
            )

    def delete_by_entity_source(self, entity_id: UUID, source: DataSource):
        with self._db_client.tx() as cursor:
            cursor.execute(
                TransactionQueries.DELETE_INVESTMENT_BY_ENTITY_SOURCE,
                (
                    str(entity_id),
                    source,
                ),
            )
            cursor.execute(
                TransactionQueries.DELETE_ACCOUNT_BY_ENTITY_SOURCE,
                (str(entity_id), source),
            )

    def get_by_id(self, tx_id: UUID) -> Optional[BaseTx]:
        with self._db_client.read() as cursor:
            cursor.execute(
                TransactionQueries.GET_INVESTMENT_BY_ID,
                (str(tx_id),),
            )
            row = cursor.fetchone()
            if row:
                return _map_investment_row(row)

            cursor.execute(
                TransactionQueries.GET_ACCOUNT_BY_ID,
                (str(tx_id),),
            )
            row = cursor.fetchone()
            if row:
                return _map_account_row(row)
        return None

    def delete_by_id(self, tx_id: UUID):
        with self._db_client.tx() as cursor:
            cursor.execute(
                TransactionQueries.DELETE_BY_ID_INVESTMENT,
                (str(tx_id),),
            )
            cursor.execute(
                TransactionQueries.DELETE_BY_ID_ACCOUNT,
                (str(tx_id),),
            )

File: mobile/src/services/database/repositories/transactionRepository.ts
import { DataManager } from "../dataManager"
import {
  BaseTx,
  AccountTx,
  StockTx,
  CryptoCurrencyTx,
  FundTx,
  FundPortfolioTx,
  BaseInvestmentTx,
  TxType,
  DataSource,
  ProductType,
  EquityType,
  FundType,
  Entity,
  TransactionQueryRequest,
  Transactions,
} from "@/domain"
import { parseDezimalValue } from "@/domain"
import { TransactionPort } from "@/application/ports"
import { TransactionQueries } from "./queries"

export class TransactionRepository implements TransactionPort {
  constructor(private client: DataManager) {}

  async save(data: Transactions): Promise<void> {
    const now = new Date().toISOString()
    await this.client.transaction(async () => {
      for (const tx of data.investment ?? []) {
        if (!tx.entity?.id || !tx.id) continue
        await this.client.execute(TransactionQueries.INSERT_INVESTMENT, {
          id: tx.id,
          ref: tx.ref,
          name: tx.name,
          amount: String(tx.amount),
          currency: tx.currency,
          type: tx.type,
          date: tx.date,
          entity_id: tx.entity.id,
          is_real: tx.source === DataSource.REAL ? 1 : 0,
          source: tx.source,
          product_type: tx.productType,
          created_at: now,
          isin: (tx as any).isin ?? null,
          ticker: (tx as any).ticker ?? (tx as any).symbol ?? null,
          asset_contract_address: (tx as any).contractAddress ?? null,
          market: (tx as any).market ?? null,
          shares: String((tx as any).shares ?? (tx as any).currencyAmount ?? 0),
          price: String((tx as any).price ?? 0),
          net_amount:
            (tx as any).netAmount != null
              ? String((tx as any).netAmount)
              : null,
          fees: String((tx as any).fees ?? 0),
          retentions:
            (tx as any).retentions != null
              ? String((tx as any).retentions)
              : null,
          order_date: (tx as any).orderDate ?? null,
          linked_tx: (tx as any).linkedTx ?? null,
          interests: (tx as any).interests ?? null,
          iban: (tx as any).iban ?? null,
          portfolio_name: (tx as any).portfolioName ?? null,
          product_subtype:
            (tx as any).equityType ?? (tx as any).fundType ?? null,
        } as any)
      }

      for (const tx of data.account ?? []) {
        if (!tx.entity?.id || !tx.id) continue
        await this.client.execute(TransactionQueries.INSERT_ACCOUNT, [
          tx.id,
          tx.ref,
          tx.name,
          String(tx.amount),
          tx.currency,
          tx.type,
          tx.date,
          tx.entity.id,
          tx.source === DataSource.REAL ? 1 : 0,
          tx.source,
          now,
          String(tx.fees ?? 0),
          String(tx.retentions ?? 0),
          tx.interestRate != null ? String(tx.interestRate) : null,
          tx.avgBalance != null ? String(tx.avgBalance) : null,
          tx.netAmount != null ? String(tx.netAmount) : null,
        ])
      }
    })
  }

  async getAll(
    real?: boolean | null,
    excludedEntities?: string[] | null,
  ): Promise<Transactions> {
    const investmentConditions: string[] = []
    const accountConditions: string[] = []
    const paramsInvestment: any[] = []
    const paramsAccount: any[] = []

    if (real != null) {
      investmentConditions.push("it.is_real = ?")
      accountConditions.push("at.is_real = ?")
      paramsInvestment.push(real ? 1 : 0)
      paramsAccount.push(real ? 1 : 0)
    }

    if (excludedEntities && excludedEntities.length > 0) {
      const placeholders = excludedEntities.map(() => "?").join(",")
      investmentConditions.push(`it.entity_id NOT IN (${placeholders})`)
      accountConditions.push(`at.entity_id NOT IN (${placeholders})`)
      paramsInvestment.push(...excludedEntities)
      paramsAccount.push(...excludedEntities)
    }

    const investmentSql =
      TransactionQueries.INVESTMENT_SELECT_BASE +
      (investmentConditions.length > 0
        ? ` WHERE ${investmentConditions.join(" AND ")}`
        : "")
    const accountSql =
      TransactionQueries.ACCOUNT_SELECT_BASE +
      (accountConditions.length > 0
        ? ` WHERE ${accountConditions.join(" AND ")}`
        : "")

    const [investment, account] = await Promise.all([
      this.client.query<any>(investmentSql, paramsInvestment),
      this.client.query<any>(accountSql, paramsAccount),
    ])

    return {
      investment: investment.rows.map(r => this.mapRowToTransaction({ ...r })),
      account: account.rows.map(r =>
        this.mapRowToTransaction({ ...r, product_type: "ACCOUNT" }),
      ) as any,
    }
  }

  async getRefsByEntity(entityId: string): Promise<any> {
    const res = await this.client.query<any>(
      TransactionQueries.GET_REFS_BY_ENTITY,
      [entityId, entityId],
    )
    return res.rows.map((r: any) => r.ref)
  }

  async getByEntity(entityId: string): Promise<Transactions> {
    const [investment, account] = await Promise.all([
      this.client.query<any>(TransactionQueries.INVESTMENT_SELECT_BY_ENTITY, [
        entityId,
      ]),
      this.client.query<any>(TransactionQueries.ACCOUNT_SELECT_BY_ENTITY, [
        entityId,
      ]),
    ])

    return {
      investment: investment.rows.map(r => this.mapRowToTransaction({ ...r })),
      account: account.rows.map(r =>
        this.mapRowToTransaction({ ...r, product_type: "ACCOUNT" }),
      ) as any,
    }
  }

  async getByEntityAndSource(
    entityId: string,
    source: DataSource,
  ): Promise<Transactions> {
    const [investment, account] = await Promise.all([
      this.client.query<any>(
        TransactionQueries.INVESTMENT_AND_ACCOUNT_BY_ENTITY_AND_SOURCE,
        [entityId, source],
      ),
      this.client.query<any>(TransactionQueries.ACCOUNT_BY_ENTITY_AND_SOURCE, [
        entityId,
        source,
      ]),
    ])

    return {
      investment: investment.rows.map(r => this.mapRowToTransaction({ ...r })),
      account: account.rows.map(r =>
        this.mapRowToTransaction({ ...r, product_type: "ACCOUNT" }),
      ) as any,
    }
  }

  async getRefsBySourceType(real: boolean): Promise<any> {
    const res = await this.client.query<any>(
      TransactionQueries.GET_REFS_BY_SOURCE_TYPE,
      [real ? 1 : 0, real ? 1 : 0],
    )
    return res.rows.map((r: any) => r.ref)
  }

  async deleteBySource(source: DataSource): Promise<void> {
    await this.client.transaction(async () => {
      await this.client.execute(
        TransactionQueries.DELETE_INVESTMENT_BY_SOURCE,
        [source],
      )
      await this.client.execute(TransactionQueries.DELETE_ACCOUNT_BY_SOURCE, [
        source,
      ])
    })
  }

  async deleteByEntitySource(
    entityId: string,
    source: DataSource,
  ): Promise<void> {
    await this.client.transaction(async () => {
      await this.client.execute(
        TransactionQueries.DELETE_INVESTMENT_BY_ENTITY_SOURCE,
        [entityId, source],
      )
      await this.client.execute(
        TransactionQueries.DELETE_ACCOUNT_BY_ENTITY_SOURCE,
        [entityId, source],
      )
    })
  }

  async getById(txId: string): Promise<BaseTx | null> {
    const [inv, acc] = await Promise.all([
      this.client.query<any>(TransactionQueries.GET_INVESTMENT_BY_ID, [txId]),
      this.client.query<any>(TransactionQueries.GET_ACCOUNT_BY_ID, [txId]),
    ])
    if (inv.rows.length > 0) return this.mapRowToTransaction(inv.rows[0])
    if (acc.rows.length > 0)
      return this.mapRowToTransaction({
        ...acc.rows[0],
        product_type: "ACCOUNT",
      })
    return null
  }

  async deleteById(txId: string): Promise<void> {
    await this.client.transaction(async () => {
      await this.client.execute(TransactionQueries.DELETE_BY_ID_INVESTMENT, [
        txId,
      ])
      await this.client.execute(TransactionQueries.DELETE_BY_ID_ACCOUNT, [txId])
    })
  }

  async getByFilters(query: TransactionQueryRequest): Promise<BaseTx[]> {
    const params: any[] = []
    let whereConditions: string[] = []

    if (query.excludedEntities && query.excludedEntities.length > 0) {
      const placeholders = query.excludedEntities.map(() => "?").join(", ")
      whereConditions.push(`tx.entity_id NOT IN (${placeholders})`)
      params.push(...query.excludedEntities)
    }

    if (query.entities && query.entities.length > 0) {
      const placeholders = query.entities.map(() => "?").join(", ")
      whereConditions.push(`tx.entity_id IN (${placeholders})`)
      params.push(...query.entities)
    }

    if (query.fromDate) {
      whereConditions.push(`tx.date >= ?`)
      params.push(query.fromDate)
    }

    if (query.toDate) {
      whereConditions.push(`tx.date <= ?`)
      params.push(query.toDate)
    }

    if (query.productTypes && query.productTypes.length > 0) {
      const placeholders = query.productTypes.map(() => "?").join(", ")
      whereConditions.push(`tx.product_type IN (${placeholders})`)
      params.push(...query.productTypes)
    }

    if (query.types && query.types.length > 0) {
      const placeholders = query.types.map(() => "?").join(", ")
      whereConditions.push(`tx.type IN (${placeholders})`)
      params.push(...query.types)
    }

    const whereClause =
      whereConditions.length > 0 ? `WHERE ${whereConditions.join(" AND ")}` : ""

    const limit = query.limit || 100
    const page = query.page || 1
    const offset = (page - 1) * limit
    params.push(limit, offset)

    const sql = `
      ${TransactionQueries.GET_BY_FILTERS_BASE}
      ${whereClause}
      ORDER BY tx.date DESC
      LIMIT ? OFFSET ?
    `

    const result = await this.client.query<any>(sql, params)
    return result.rows.map(this.mapRowToTransaction)
  }

  async getRecentTransactions(limit: number = 10): Promise<BaseTx[]> {
    const sql = `
      ${TransactionQueries.GET_BY_FILTERS_BASE}
      ORDER BY tx.date DESC
      LIMIT ?
    `

    const result = await this.client.query<any>(sql, [limit])
    return result.rows.map(this.mapRowToTransaction)
  }

  async getTransactionsByEntity(
    entityId: string,
    limit: number = 50,
  ): Promise<BaseTx[]> {
    const sql = `
      ${TransactionQueries.GET_BY_FILTERS_BASE}
      WHERE tx.entity_id = ?
      ORDER BY tx.date DESC
      LIMIT ?
    `
    const result = await this.client.query<any>(sql, [entityId, limit])
    return result.rows.map(this.mapRowToTransaction)
  }

  async getTransactionsByDateRange(
    startDate: Date,
    endDate: Date,
    limit: number = 100,
  ): Promise<BaseTx[]> {
    const s = startDate.toISOString()
    const e = endDate.toISOString()
    const sql = `
      ${TransactionQueries.GET_BY_FILTERS_BASE}
      WHERE tx.date >= ? AND tx.date <= ?
      ORDER BY tx.date DESC
      LIMIT ?
    `
    const result = await this.client.query<any>(sql, [s, e, limit])
    return result.rows.map(this.mapRowToTransaction)
  }

  private mapRowToTransaction(row: any): BaseTx {
    const entity: Entity = {
      id: row.entity_id,
      name: row.entity_name,
      naturalId: row.entity_natural_id ?? null,
      type: row.entity_type,
      origin: row.entity_origin,
      iconUrl: row.icon_url,
    }

    const base: BaseTx = {
      id: row.id,
      ref: row.ref,
      name: row.name,
      amount: parseDezimalValue(row.amount),
      currency: row.currency,
      type: row.type as TxType,
      date: row.date,
      entity: entity,
      source: row.source as DataSource,
      productType: row.product_type as ProductType,
    }

    if (row.product_type === "ACCOUNT") {
      return {
        ...base,
        fees: parseDezimalValue(row.fees),
        retentions: parseDezimalValue(row.retentions),
        interestRate:
          row.interest_rate != null
            ? parseDezimalValue(row.interest_rate)
            : null,
        avgBalance:
          row.avg_balance != null ? parseDezimalValue(row.avg_balance) : null,
        netAmount:
          row.net_amount != null ? parseDezimalValue(row.net_amount) : null,
      } as AccountTx
    }

    // Investment Transactions
    const investmentBase: BaseInvestmentTx = { ...base }

    if (row.product_type === ProductType.STOCK_ETF) {
      return {
        ...investmentBase,
        shares: parseDezimalValue(row.shares),
        price: parseDezimalValue(row.price),
        fees: parseDezimalValue(row.fees),
        netAmount:
          row.net_amount != null ? parseDezimalValue(row.net_amount) : null,
        isin: row.isin,
        ticker: row.ticker,
        market: row.market,
        retentions:
          row.retentions != null ? parseDezimalValue(row.retentions) : null,
        orderDate: row.order_date,
        linkedTx: row.linked_tx,
        equityType: row.product_subtype as EquityType,
      } as StockTx
    } else if (row.product_type === ProductType.CRYPTO) {
      return {
        ...investmentBase,
        currencyAmount: parseDezimalValue(row.shares), // mapped from shares
        symbol: row.ticker, // mapped from ticker
        price: parseDezimalValue(row.price),
        fees: parseDezimalValue(row.fees),
        contractAddress: row.asset_contract_address,
        netAmount:
          row.net_amount != null ? parseDezimalValue(row.net_amount) : null,
        retentions:
          row.retentions != null ? parseDezimalValue(row.retentions) : null,
        orderDate: row.order_date,
      } as CryptoCurrencyTx
    } else if (row.product_type === ProductType.FUND) {
      return {
        ...investmentBase,
        shares: parseDezimalValue(row.shares),
        price: parseDezimalValue(row.price),
        fees: parseDezimalValue(row.fees),
        netAmount:
          row.net_amount != null ? parseDezimalValue(row.net_amount) : null,
        isin: row.isin,
        market: row.market,
        retentions:
          row.retentions != null ? parseDezimalValue(row.retentions) : null,
        orderDate: row.order_date,
        fundType: row.product_subtype as FundType,
      } as FundTx
    } else if (row.product_type === ProductType.FUND_PORTFOLIO) {
      return {
        ...investmentBase,
        portfolioName: row.portfolio_name,
        iban: row.iban,
        fees: row.fees != null ? parseDezimalValue(row.fees) : undefined,
      } as FundPortfolioTx
    }

    // Default fallback for other investment types
    return investmentBase
  }
}
